storage/test/sparse_merkle/tree_test.cpp

```diff
@@ -273,13 +406,658 @@ TEST(tree_tests, multiple_updates) {
   ASSERT_EQ(2, batch.stale.internal_keys.size());
   ASSERT_EQ(0, batch.stale.leaf_keys.size());
 
-  // The paths of the stale node keys should be "" and "b"
-  ASSERT_EQ("", batch.stale.internal_keys[0].path().toString());
-  ASSERT_EQ("b", batch.stale.internal_keys[1].path().toString());
+  // The paths of the stale node keys should be "b" and "" (bottom up order),
+  // and they should exist at version 3.
+  ASSERT_TRUE(internalKeyExists("b", 3, batch.stale.internal_keys));
+  ASSERT_TRUE(internalKeyExists("", 3, batch.stale.internal_keys));
+}
+
+// Insert 2 keys at the root and remove one of them in a second update. The
+// remove should return a stale root node and an updated root node.
+TEST(tree_tests, remove_complete_at_root_test) {
+  std::shared_ptr<TestDB> db(new TestDB);
+  Tree tree(db);
+  SetOfKeyValuePairs updates;
+  updates.emplace(Sliver("key1"), Sliver("val1"));
+  updates.emplace(Sliver("key2"), Sliver("val2"));
+  auto batch = tree.update(updates);
+  // There should be two leafs in the root
+  ASSERT_EQ(2, batch.internal_nodes[0].second.numLeafChildren());
+  db_put(db, batch);
+
+  // Delete one of the children
+  KeysVector deletes{Sliver("key1")};
+  batch = tree.remove(deletes);
+  db_put(db, batch);
+
+  // The root should only have one leaf child, and should be the only internal
+  // node that needs writing.
+  ASSERT_EQ(1, batch.internal_nodes.size());
+  ASSERT_EQ(0, batch.leaf_nodes.size());
+  auto& [root_key, root_node] = batch.internal_nodes[0];
+  ASSERT_EQ(Version(2), root_key.version());
+  ASSERT_EQ(Version(2), root_node.version());
+  ASSERT_EQ(1, root_node.numLeafChildren());
+
+  // The old root node is the only stale internal node. The deleted LeafKey at
+  // the version 1 should match the hash of "key1".
+  ASSERT_EQ(Version(2), batch.stale.stale_since_version);
+  ASSERT_EQ(1, batch.stale.internal_keys.size());
+  ASSERT_TRUE(internalKeyExists("", 1, batch.stale.internal_keys));
+  ASSERT_EQ(1, batch.stale.leaf_keys.size());
+  ASSERT_TRUE(leafKeyExists("key1", 1, batch.stale.leaf_keys));
+}
+
+// Insert 2 keys that match in the first Nibble, and are inserted at the same
+// InternalNode at depth 1. Deleting one of them should cause the other one to
+// be promoted, and reside in the root. The root is the only new InternalNodeKey
+// to be written.
+TEST(tree_tests, remove_and_promote) {
+  std::shared_ptr<TestDB> db(new TestDB);
+  Tree tree(db);
+  SetOfKeyValuePairs updates;
+  updates.emplace(Sliver("key3"), Sliver("val1"));
+  updates.emplace(Sliver("key9"), Sliver("val2"));
+  auto batch = tree.update(updates);
+
+  // There should be 2 internal nodes, with no leaf children in the root node.
+  ASSERT_EQ(2, batch.internal_nodes.size());
+  auto& root_node = batch.internal_nodes[0].second;
```

Comment:
Can we rely on internal node ordering in the batch? If yes, maybe we can document it. Or maybe we can encourage users to use the root getters in `Tree`. I realise this is test code and my comment is mostly about the API.

Reply:
We actually can rely on the ordering at this moment, but I'm not sure we should. We may end up changing this to be a set instead of a vector, although a vector probably has better perf characteristics (if it matters.). It's mainly just internal knowledge used in testing. I prefer to leave it undocumented for now. Of course, if we change it, then we'll need to rework the test.

Reply:
Agreed.